dvc/project.py

```diff
@@ -1165,22 +1157,33 @@ def metrics_add(self, path, typ=None, xpath=None):
     def metrics_remove(self, path):
         self._metrics_modify(path, delete=True)
 
-    def graph(self, from_directory=None):
+    def graph(self, stages=None, from_directory=None, cwd=None):
         import networkx as nx
-        from dvc.exceptions import OutputDuplicationError
+        from dvc.exceptions import (OutputDuplicationError,
+                                    WorkingDirectoryAsOutputError)
 
         G = nx.DiGraph()
         G_active = nx.DiGraph()
-        stages = self.stages(from_directory)
-
+        stages = stages or self.stages(from_directory)
+        stages = [stage for stage in stages if stage]
         outs = []
         outs_by_path = {}
+
+        if cwd:
+            cwd_path = os.path.abspath(os.path.normpath(cwd))
+        else:
+            cwd_path = None
+
         for stage in stages:
             for o in stage.outs:
-                existing = outs_by_path.get(o.path, None)
-                if existing is not None:
+                existing = outs_by_path.get(o.path)
+                if existing:
                     stages = [o.stage.relpath, existing.stage.relpath]
                     raise OutputDuplicationError(o.path, stages)
+
+                if cwd_path and o.path == cwd_path:
```

Comment:
https://github.com/iterative/dvc/pull/1460#discussion_r247386225

We only check it for the specified cwd, which only relates to the current stage that we are creating. And it is only enabled for `dvc run` right now. What we should do instead, is to check that every stage in the DAG doesn't have cwd that overlaps with outputs of other stages.

While collecting stages in Project.stages(), we are able to simply skip walking through the directories that are outputs to the already collected stages. This covers us for cases when a stage has `./out` path to the output and doesn't help us when it has something weird like `../../out`, because we are walking the project's tree top-down. So we need to check the whole DAG for such errors.

Reply:
So we are currently covering:
```
dvc run -o data/ 'rm -rf data && mkdir data'
dvc run --cwd data/ 'echo something > something'
```

However, we are missing:
```
dvc run -o data/ 'rm -rf data && mkdir data'

cat <<YAML > data/something.dvc
cmd: echo something > something
outs:
- path: something
YAML

dvc repro data/something.dvc
```

Is that what you are refering too, @efiop ?

Would you mind writting a script with the case we are not covering?

_NOTE: The later one is raising an `unexpected error`_

Thanks for your patience, it has been really helpful for me :)

Reply:
In both cases, it should actually not see `something.dvc` at all, because while collecting dvcfiles, it will see that `data/` is an output of `data.dvc` and will simply skip scanning it. See Project.stages(). I do however see `Unexpected error` as well, so it is definitely related. Need to investigate what went wrong.

What I actually meant is little bit more tricky because of the scanning optimization mentioned above. Here is your second script modified:

```
#!/bin/bash

set -e
set -x

rm -rf myrepo
mkdir myrepo

cd myrepo

git init
dvc init

mkdir dir1
dvc run -c dir1 -o ../dir2 'rm -rf ../dir2 && mkdir ../dir2'

cat <<YAML > dir2/something.dvc
cmd: echo something > something
outs:
- path: something
YAML

dvc repro -v dir2/something.dvc
```

As you can see, even though `something.dvc` is inside a directory that was created by the first stage, `dvc repro` still succeeds, even though it should've raised an error. This happens because we don't check that all stages don't overlap with outputs of other stages. Also, please note that checking `o.path == cwd` is not enough, we should rather check that cwd doesn't start with `o.path + os.sep`(e.g. cwd == dir/subdir/, and o.path is dir/).